smarts/waymo/waymo_data_extraction.py

```diff
@@ -0,0 +1,310 @@
+        rgb=True,
+        road_waypoints=False,
+        waypoints=False,
+    ),
+)
+
+
+def _scenario_code(dataset_path, scenario_id):
+    return f"""from pathlib import Path
+
+from smarts.sstudio import gen_scenario
+from smarts.sstudio import types as t
+
+dataset_path = "{dataset_path}"
+scenario_id = "{scenario_id}"
+
+traffic_histories = [
+    t.TrafficHistoryDataset(
+        name=f"waymo",
+        source_type="Waymo",
+        input_path=dataset_path,
+        scenario_id=scenario_id,
+    )
+]
+
+gen_scenario(
+    t.Scenario(
+        map_spec=t.MapSpec(
+            source=f"{{dataset_path}}#{{scenario_id}}", lanepoint_spacing=1.0
+        ),
+        traffic_histories=traffic_histories,
+    ),
+    output_dir=Path(__file__).parent,
+)
+"""
+
+
+def _record_data(
+    smarts,
+    collected_data,
+    output_dir,
+    current_vehicles,
+    off_road_vehicles,
+    selected_vehicles,
+    std_obs_wrapper,
+):
+    # Attach sensors to each vehicle
+    valid_vehicles = (current_vehicles - off_road_vehicles) & selected_vehicles
+    for veh_id in valid_vehicles:
+        try:
+            smarts.attach_sensors_to_vehicles(AGENT_SPEC.interface, {veh_id})
+        except ControllerOutOfLaneException:
+            logger.warning(f"{veh_id} out of lane, skipped attaching sensors")
+            off_road_vehicles.add(veh_id)
+
+    # Get observations from each vehicle and record them
+    obs: Dict[str, Observation] = {}
+    obs, _, _, _ = smarts.observe_from(list(valid_vehicles))
+    resolutions = {}
+    for id_ in list(obs):
+        resolutions[id_] = obs[id_].top_down_rgb.metadata.resolution
+        if obs[id_].ego_vehicle_state.lane_index == None:
+            del obs[id_]
+    obs = std_obs_wrapper.observation(obs)
+    t = smarts.elapsed_sim_time
+    for car, car_obs in obs.items():
+        collected_data.setdefault(car, {}).setdefault(t, {})
+        collected_data[car][t] = car_obs
+
+    # Write top-down RGB image to a file for each vehicle
+    for agent_id, agent_obs in obs.items():
+        h, w = agent_obs["rgb"].shape[0], agent_obs["rgb"].shape[1]
+        shape = [
+            (h / 2 - 1.47 / 2 / resolutions[id_], w / 2 - 3.68 / 2 / resolutions[id_]),
+            (h / 2 + 1.47 / 2 / resolutions[id_], w / 2 + 3.68 / 2 / resolutions[id_]),
+        ]
+        img = Image.fromarray(agent_obs["rgb"], "RGB")
+        rect_image = ImageDraw.Draw(img)
+        rect_image.rectangle(shape, fill="red")
+        img.save(output_dir / f"{t}_{agent_id}.png")
+
+
+def main(
+    dataset_file: Path,
+    scenario_id: str,
+    output_dir: Path,
+    scenarios_dir: Path,
+    vehicle_ids: Optional[List[int]] = None,
+    max_time: float = 20.0,
+):
+    """Extract top-down RGB images and observations from a Waymo Motion Dataset scenario.
+    Args:
+        dataset_file (Path):
+            Path to the TFRecord file.
+        scenario_id (str):
+            ID for the scenario in the TFRecord file.
```

Comment:
Can we extract Waymo data into Smarts observations for multiple scenarios in one go?